unix_integration/pam_kanidm/src/pam/mod.rs

```diff
@@ -309,6 +312,61 @@ impl PamHooks for PamKanidm {
                         );
                         continue;
                     }
+                    ClientResponse::PamAuthenticateStepResponse(PamAuthResponse::MFACode {
+                        msg,
+                        data,
+                    }) => {
+                        let cred = match conv.send(PAM_PROMPT_ECHO_OFF, &msg) {
+                            Ok(password) => match password {
+                                Some(cred) => cred,
+                                None => {
+                                    debug!("no mfa code");
+                                    return PamResultCode::PAM_CRED_INSUFFICIENT;
+                                }
+                            },
+                            Err(err) => {
+                                debug!("unable to get mfa code");
+                                return err;
+                            }
+                        };
+
+                        // Now setup the request for the next loop.
+                        timeout = cfg.unix_sock_timeout;
+                        req = ClientRequest::PamAuthenticateStep(PamAuthRequest::MFACode {
+                            cred,
+                            data,
+                        });
+                        continue;
+                    }
+                    ClientResponse::PamAuthenticateStepResponse(PamAuthResponse::MFAPoll {
+                        msg,
+                        max_poll_attempts,
```

Comment:
Shouldn't the fact we are in a poll state be in the auth session? And it it tracks this? 

Said better, the resolver is controlling the whole flow. The pam module is just an information pipe. So the resolver when it presents the challenge to pam along with the poll interval. then pam module implicitly moves to "start polling". Then pam module just sends a "poll" message with no options. 

It's up to the resolver to track the timeout in the authsession, and indicate that the pam module continues to wait, errors, or has succeeded etc. 

To me I'd assume the PamAuthRequest::MFAPoll shouldn't have any optinos at all, and when we start the poll we tell the client in the clientResponse only the polling interval. 

It's up to the resolvers auth_session state to track the number of polls etc.

Does that seem like it could be a better approach?

Reply:
I guess what I'm saying is the pam module should have less logic, it's just polling the resolver until the resolver decides and tells it what to do.

Reply:
Ok, I see what you mean. Looking into it.

Reply:
Well, I've tried today, but I'm having trouble wrapping my head around it. I suppose the auth provider would start the polling, but where would the auth_session init happen (a polling init)? This seems like an awful lot of complexity for something so simple.

Reply:
Maybe think of it like this. The pam module is just asking "are we there yet" like a kid on a road trip. 

In the background the auth_session has spawn an extra task that does the communication to the provider in it's own internal loop and on completion can update a status that's behind an async mutex in the auth session.

And yeah, it seems complex, but we are dealing with multiple moving parts (pam, resolver and azure ad) as well as a human. The human can ctrl c things, network errors, or anything else. So we have to model it as an event driven system and have states that are transitioned through. 

And the reason we add this now in this way, is that we can re-use the same process and parts later in other things because what we are really modeling is a generic-concurrent auth checker.

WRT to the init, it's the same as any interaction with pam. The session starts https://github.com/kanidm/kanidm/blob/master/unix_integration/src/resolver.rs#L867 and then we have next_req. So the auth_session can immediately make this `PamAuthResponse::BeginPollWait { msg: String, interval: u8 }`. In the background auth_session has already started the task to communicate and poll itself to azure etc. 

Then pam sets "PamAuthRequest::Poll". The auth session + resolver checks their async mutex about the task state and sees it's still "Waiting" and would return PamAuthResponse::PollWait.

If an error happens then the response is PamAuthResponse::Denied. 

If pam requests PamAuthRequest::Poll and the auth session + resolver is ready and the status is "denied, set Denied in the response. If it's "Success" then you can make the next choice to either return PamAuthResponse::Success or you can then start something like PamAuthResponse::PinEnroll to get the user to set their PIN on first use or something.

Reply:
> Maybe think of it like this. The pam module is just asking "are we there yet" like a kid on a road trip.

I get the broader picture. I'm having trouble executing your vision of it.

> WRT to the init, it's the same as any interaction with pam. The session starts https://github.com/kanidm/kanidm/blob/master/unix_integration/src/resolver.rs#L867 and then we have next_req. So the auth_session can immediately make this `PamAuthResponse::BeginPollWait { msg: String, interval: u8 }`. In the background auth_session has already started the task to communicate and poll itself to azure etc.

In himmelblau, I don't even know at this point that I'll be doing a poll (when in `pam_account_authenticate_init()`). That doesn't happen until after a response from `unix_user_online_auth_step()`. This is because I first send just a username+password req to Azure, and if it responds with an error saying MFA is required, then we start that dance.

> Then pam sets "PamAuthRequest::Poll". The auth session + resolver checks their async mutex about the task state and sees it's still "Waiting" and would return PamAuthResponse::PollWait.

I also don't see the purpose of leaving a task polling here. This is a single call every X # of seconds (as specified by Azure). IMHO, we should just wait in PAM, sending a request to the daemon every so often (the interval specified by Azure), until the daemon responds with an auth success.
Why spin a daemon to poll Azure, when all it's going to do is sit there waiting for us to send a request from PAM for the auth status?

Reply:
![MFA_challenge drawio](https://github.com/kanidm/kanidm/assets/2014292/5624addf-402b-41da-bc52-e22ba35761d5)

This is what it's currently doing. The resolver should only send a poll status request to Azure if PAM asks for it, IMHO. There's no reason for the resolver to be spinning and polling simultaneously as PAM is spinning and sending status requests to the resolver.

Currently my 'wait' response is just sending a new `PamAuthResponse::MFAPoll` to PAM. This isn't necessary, but was just a sort of simplification (that's the reason for the extra data, is I just combined it into a single request). Perhaps we could have a sub loop within the initial `PamAuthResponse::MFAPoll` that doesn't require passing the same data over and over?

Reply:
> Currently my 'wait' response is just sending a new `PamAuthResponse::MFAPoll` to PAM. This isn't necessary, but was just a sort of simplification (that's the reason for the extra data, is I just combined it into a single request). Perhaps we could have a sub loop within the initial `PamAuthResponse::MFAPoll` that doesn't require passing the same data over and over?

I implemented this as a sub-loop in the latest changes here.

Reply:
And of course, then you don't need "max attempts" or counting or anything. Pam just keeps polling until the resolver says "no" . Then it's completely up to the resolver and it's timeouts and stuff with it's session to azure ad which influences the whole thing. The logic now firmly becomes tied only to the resolver and azure communication and it's timeouts, rather than multiple possible timeouts that can occur.

Reply:
The other problem is with adjusting the timeout like https://github.com/kanidm/kanidm/blob/master/unix_integration/pam_kanidm/src/pam/mod.rs#L306 then this taints the timeout for the rest of the pam module session where it's now possible to leave the user waiting for a really long time if the resolver hasn't returned.

Reply:
for comparison, what do you think are the benefits to the way you wanted to do it? Now I'm getting in my own head about if my ideas are right or not.

Reply:
I'm getting in my head now about if I was right in what I said at all. Maybe there is an easier way, I'm not sure. I feel like I've sent you on a wild chase now. 

Possibly the right answer is in front of us: 

https://github.com/kanidm/kanidm/blob/master/unix_integration/pam_kanidm/src/pam/mod.rs#L244

pam here has a timeout so that if the resolver takes too long for an expected action, it fails. In my head I was thinking this timeout detects disconnects of the resolver, but that's handled by the err case in the loop. 

When the resolver says "wait" we tell pam an upper limit to wait for. Then the resolver can do it's work and no polling needed, and when the resolver is done it just sends back a response. 

I feel extremely stupid not seeing the simple and obvious answer here. I think we don't want polls at all. 


<img width="940" alt="Screenshot 2024-03-15 at 12 21 06" src="https://github.com/kanidm/kanidm/assets/271005/2549c41a-e105-4167-b862-d217e950c1a2">

Pam calls the resolver, which sends back "present the mfa challenge to the user". When it's displayed to the user, pam messages us to say "yep, they've seen it". That now sets pam blocking on the socket to wait for a response message. In the "present challenge" we send a "long timeout" adjustment.

Then the resolver can sit there and poll on it's own schedule as needed. Internally the resolver should have a "proper" timeout incase things take too long. 

This way the resolver then only needs to send back one message wen the communication is complete, and then pam can advance as well. 

I think this is way simpler, doesn't need any polling, and will be better for you to implement, and covers our needs. 

What do you think? I feel really really really bad giving you such a run around on this, I'm really sorry.

Reply:
Well, that's how the DAG works right now in there. I assumed you meant you wanted that redesigned to do polling as well (I was going to do this next).

Reply:
So this whole conversation wasn't a waste. I actually am going to argue in favor of polling.

The way the current patches are written, PAM is waiting in between sending requests to the resolver. The resolver does nothing unless prompted to check the status.

If we do a long wait instead, PAM is waiting on the response, and the resolver is spinning and calling Azure periodically.

Either way will work, but think about what happens when a user terminates the authentication attempt. If PAM is just periodically requesting an update, then the whole session immediately terminates (because the resolver wasn't doing anything anyway, unless PAM asked it to). If pam is waiting instead on a timeout, then the resolver **doesn't stop asking** Azure for an update until the max tries have completed.

So, consider the implications if say 1000 users did the same thing... started to authenticate, but bailed out all at once. Now we've got our resolver bogged down with many requests, none of which will terminate until we reach **max_poll_attempts * polling_interval**, and none of which PAM cares to see completed anyway. This doesn't scale well.

Reply:
This is a very valid and legitimate concern, but there is a way to handle this, and it may be cleaner than we expect.

If the pam session terminates then https://github.com/kanidm/kanidm/blob/master/unix_integration/src/daemon.rs#L206 drops and yields None, going to https://github.com/kanidm/kanidm/blob/master/unix_integration/src/daemon.rs#L410

We can use this behaviour to then signal that the session dropped to the underlying resolver. All we have to do is add https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html 

We setup the broadcast at the start, and then for requests that need it, we send the subscriber side to the cachelayer such as `cachelayer.pam_account_authenticate_step(auth_session, pam_next_req, liveness_subscriber)`. 

Then within the authenticate step, if we ever wait on something (such as our loop for polling azure) we do it in a select such as:

```
while max_poll > 0 {
    max_poll--;
select! {
    Ok(something) = azure.poll() => {}
   _ = timeout => { continue }
    _ = liveness_subscriber => { break }
}
```

The trick with the liveness_subscriber is any message on it *or* if the broadcast sender drops, it causes an event on the subscriber allowing immediate cancellation. 

Then around line 410 in daemon we just have the broadcast.send(()) just to cause the subscriber to activate and cancel.

Does that seem like a good option? It means we keep all the max poll logic and timeouts in the resolver (himmleblau) while also having a way to signal to spawned tasks that they have been canceled.

Reply:
That sounds ok. I've been pulled away to Samba code for a bit, but I'll get back to this soon.

Reply:
Turns out it won't work anyway. After sleeping on it, you are correct that we need polling. The reason is that the outer daemon loop won't be on the recv().await, it will be awaiting the inner resolver task.

So we do need polling. 

I suggest we go back to what I suggested before, where we instruct pam to just poll in a loop until a result is ready. All the timeouts and logic for max attempts/polls should be in the resolver (not pam). and then we can use the same broadcast channel idea to solve notifying the resolver that the pam client has disconnected. How does that sound?

Reply:
I have no idea how to implement what you're suggesting. This has gotten far too complex for something so straight forward and simple. sigh

Reply:
I'm going to push back and say I still think polling on the resolver is the wrong approach. I think the polling should happen in PAM. This makes the whole thing very straight forward and simple. This would mean PAM asks every so often for an update, meanwhile the resolver does nothing until getting a explicit request from the resolver.
Why do you want the polling to happen in the resolver? I still don't think you've given a good reason.

Reply:
> I have no idea how to implement what you're suggesting. This has gotten far too complex for something so straight forward and simple. sigh

We have an event driven system, it's never simple. 

I'll update the PR then.